### ISSUE TITLE
[FR] Support skops as model storage format

### ISSUE THREAD

**Issue by lorentzenchr** [mlflow/mlflow#13077]
### Willingness to contribute

No. I cannot contribute this feature at this time.

### Proposal Summary

Consider to support https://github.com/skops-dev/skops (see `skops.io`) as storage format for models.

### Motivation

> #### What is the use case for this feature?
Instead of pickle, use skops format to store trained models permanently.

> #### Why is this use case valuable to support for MLflow users in general?
Skops is a much safer alternative to pickle, see https://skops.readthedocs.io/en/latest/persistence.html.

> #### Why is this use case valuable to support for your project(s) or organization?
Security concerns, e.g. about code execution with pickle files.

> #### Why is it currently difficult to achieve this use case?
There is hardly any alternative available to pickle files for storing trained models in python (apart from the https://mlflow.org/docs/latest/models.html#models-from-code).

### Details

@adrinjalali and @BenjaminBossan (both skops developers) friendly ping

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by adrinjalali** [comment]
I'd be more than happy to give a hand if it's decided that it'd be nice to support `skops.io` formats.

**Comment by lorentzenchr** [comment]
@BenWilson2 @dbczumar @daniellok-db @gabrielfu @harupy @serena-ruan @B-Step62 @TomeHirata friendly ping. Any advice as to how to proceed to come to a discussion and decision would be very appreciated.

**Comment by adrinjalali** [comment]
Just an update that I've been working for a couple of weeks on making `skops.io` more stable, remove a bunch of dependencies, and making the persistence more "production ready" in a sense, and my plan is to work on this plugin after that stage, which should come in the coming weeks.

I've studied the mlflow code a little bit when it comes to "flavors", and I might need some help when I get to it, and that help would be very much appreciated.

**Comment by lorentzenchr** [comment]
@adrinjalali fyi: Maybe joining the mlflow office hour, see https://lu.ma/mlflow, would help to get a reaction. But I don't have the time for it.

**Comment by lorentzenchr** [comment]
@BenWilson2 @dbczumar @daniellok-db @gabrielfu @harupy @serena-ruan @TomeHirata @WeichenXu123 @B-Step62 Dear core members, it would be very nice and appreciated if you could provide any response to this feature request. With respect to security, I think @adrinjalali's solution with skops is by far superior to pickle and - after all - scikit-learn is a big part of ML.
Could you at least provide a hint as to where or how to contact you in an appropriate way?

**Comment by lorentzenchr** [comment]
I tried the slack channel: https://mlflow-users.slack.com/archives/CBFQ95Y2G/p1755520142629219

**Comment by TomeHirata** [comment]
Hi, @lorentzenchr. Thank you for the feature request. Can I confirm skops cannot be used for arbitrary python functions? We could use skops for some flavors like sklearn, but the inconsistency of model serialization library is a bit concerning. We commonly use cloudpickle now.

**Comment by lorentzenchr** [comment]
@TomeHirata Thanks for participating in this issue. I let @adrinjalali answer your question, he knows better than me.

**Comment by adrinjalali** [comment]
> Can I confirm skops cannot be used for arbitrary python functions?

Correct. It has been in our roadmap, but we haven't had a request from users for this and therefore it hasn't moved forward. It's somewhat contrary to the security `skops` persistence provides, since the user would need to read the source code of a function to see if it's safe to load or not.

> but the inconsistency of model serialization library is a bit concerning

@TomeHirata do you mean `skops` is inconsistent, or that being inconsistent in terms of not using the same library for many flavors? I'm not sure if I understand this point.

> We commonly use cloudpickle now.

Interesting. However, `cloudpickle` itself discourages this practice though ([link](https://github.com/cloudpipe/cloudpickle)):

> Using `cloudpickle` for __long-term object storage is not supported and strongly discouraged__.

In general, in this case convenience and security work somewhat against one another, and from what we've seen, the existing `skops` features support majority of use cases out there. That said, I don't think we're asking to have it as the default flavor, but rather as a possible flavor for users who wish to use it.

**Comment by TomeHirata** [comment]
> do you mean skops is inconsistent, or that being inconsistent in terms of not using the same library for many flavors? I'm not sure if I understand this point.

The latter, having different model serialization methods for different flavors would make the MLflow dependency larger and complicate the logic as we need to deal with multiple file types. Currently, we're using cloudpickle commonly for any flavors that simplifies out interface/logic. It would be great if skops could covert the arbitrary python objects like `cloudpickle` while I understand it contradicts how `skops` offers security. 

> That said, I don't think we're asking to have it as the default flavor, but rather as a possible flavor for users who wish to use it.

As we sometimes received security concerns about the pickle usage, I think having an option to use skops for possible flavors may make sense to improve the security for some usecase. You mentioned that you would be able to contribute to this feature previously, does it remain the same? I can discuss with the team if this is the case (no promise this FR will be accepted though).

**Comment by adrinjalali** [comment]
> having different model serialization methods for different flavors would make the MLflow dependency larger

Would a soft dependency be an option? With a nice error message to users to install `skops` themselves if they want to use this feature?

> complicate the logic as we need to deal with multiple file types.

It's almost common practice in the pydata/ml space to have two types to persists though. Numpy has pickle vs numpy format, keras does pickle or h5(optional dependency), and others might use `safetensors` as an alternative to pytorch's model which does (or at least used to) have pickle inside. One potential path forward, would be to allow one/two additional args when adding/loading a model artifact to let users use the alternative file type. And we can certainly start by having it in one flavor and see how the code looks like and how much complexity it adds?

> It would be great if skops could covert the arbitrary python objects like `cloudpickle`

We can certainly work on it. Developers of `cloudpickle` are also my colleagues, and I'm sure we can find a way. But it would be nice to know how often it happens that one needs them in reality.

> You mentioned that you would be able to contribute to this feature previously, does it remain the same?

Happy to.

**Comment by TomeHirata** [comment]
Thanks @adrinjalali. We're happy to review your PR.
